design-book/src/design-constraints.md

```diff
@@ -0,0 +1,53 @@
+
+Below each bullet point, motivation is given.
+
+## Must
+
+- The editor must target the latest stable Bevy release.
+  - This eases project management by separating the work across two repositories, and ensures that we aggressively test each Bevy release's migration path.
+- The Bevy editor must be a Bevy application.
+  - This is essential both for dogfooding and to ease development: Bevy contributors all know how to write Bevy!
+
+## Should
+
+- Reusable code (camera controllers, inspectors, UI widgets...) should be cleanly separated into dedicated, documented libraries.
+  - This keeps our repository clean, and allows users to steal our work to use in their own projects and tooling.
+- Widgets, styles, colors and interaction paradigms should be consistent across the editor.
+  - This avoids duplication, makes changes easier and helps give the editor a cohesive look.
+- The editor should be distributed as an executable, either directly from the Bevy website or via Steam.
+  - This makes it easier for non-technical users to install.
+- The editor should be configurable, and easily adapted to both different genres and specific games.
+  - Every game needs a different set of features. We need to allow teams to expose or extend what they care about, and hide what they don't.
+- Industry-standard conventions for controls, UX and so on should be followed unless there's a good reason not to.
+  - Learning a new tool is hard enough: please don't make it any harder.
+- The Bevy editor should be stable, and recover from crashes in the game or the editor as gracefully as possible.
+  - Losing work sucks a lot!
+- Iteration times should be kept fast, using whatever means make sense for a given workflow.
+  - Compiling Rust is slow, and more iterations speeds up work and leads to a better result.
+- Version-controllable: assets and settings should be feasible to check into version control and be relatively robust to merge conflicts.
+  - This makes sharing settings across teams easier, and reduces confusing and wasteful problems with consolodating changes.
+
+## Should not
+
+- 3rd party Bevy dependencies should be used sparingly, and only with the consent of the maintainer.
+  - Bevy's ecosystem is one of its strengths, and we should lean on it, but being incorporated into the editor is a huge maintenance burden.
+- The editor is not an asset or code creation tool (certainly not in the default configuration). If an established tool does a good job authoring content, focus on importing its output, not reinventing Blender, VSCode Asesprite and Reaper.
+  - We have limited resources, and this isn't a good use of them. Artists and professionals already know and prefer specialized tools.
+- The Bevy editor should not have a native look-and-feel.
+  - This is a poor use of resources, and makes it harder to test across platforms. Instead, focus on the useful behavior and design conventions.
+- End users should not have to install or update Rust.
```

Comment:
But why? They need Rust to run the game anyways, don't they?

Reply:
You can ship your teams binaries. That's what the C++ teams do!

Reply:
How does this look in a version controlled project? Does the project repo store the latest binaries? Does CI build the binaries and have repo write access? Are binaries archived somewhere else?

Regarding C++ teams, doesn't Unreal require a bunch of MS C++ tooling to be installed alongside it?

Reply:
If the rust toolchain is needed, it needs to be part of the editor installation flow. When I tried to get our artist to install rustup, I first had to explain to him what the terminal was. :smile:

Reply:
The workflow here is really clear: One person on a team must have rustc, and must be able to compile editor/game binaries for the rest of the team. Rust's cross completion is great, lets take advantage of it.

Deploying a rust app is so much easier than setting up a development environment on someone else's machine. 

> But why? They need Rust to run the game anyways, don't they?

Are you imagining that player will install rust when they buy your game?

Reply:
> Are you imagining that player will install rust when they buy your game?

You know what I meant :P
Alright then, if the distribution workflow is easier to do than to setup `rustup` through a UI, then that sounds like a good plan :)